### ISSUE TITLE
`needless_pass_by_ref_mut` false positive in closure in async

### ISSUE THREAD

**Issue by mkroening** [rust-lang/rust-clippy#11380]
### Summary

Another variant of https://github.com/rust-lang/rust-clippy/issues/11179 and https://github.com/rust-lang/rust-clippy/issues/11299.

Found here: [hermit-os/kernel/src/fd/socket/tcp.rs#L69-L94](https://github.com/hermit-os/kernel/blob/ca10aa9d7291aab4fe8e3e77f88db6f44f63ce45/src/fd/socket/tcp.rs#L69-L94)

### Lint Name

needless_pass_by_ref_mut

### Reproducer

I tried this code:

```rust
async fn foo(x: &mut i32) {
    (|| *x += 1)();
}
```

I saw this happen:

```
warning: this argument is a mutable reference, but not used mutably
 --> foo.rs:1:17
  |
1 | async fn foo(x: &mut i32) {
  |                 ^^^^^^^^ help: consider changing to: `&i32`
  |
  = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#needless_pass_by_ref_mut
  = note: `#[warn(clippy::needless_pass_by_ref_mut)]` on by default
```

I expected to see this happen:

No `needless_pass_by_ref_mut`.

### Version

```text
rustc 1.74.0-nightly (ef85656a1 2023-08-21)
binary: rustc
commit-hash: ef85656a10657ba5e4f7fe2931a4ca6293138d51
commit-date: 2023-08-21
host: aarch64-apple-darwin
release: 1.74.0-nightly
LLVM version: 17.0.0
```

**Comment by GuillaumeGomez** [comment]
It'd be better indeed. I started looking at it but I'm a bit under the water currently, too many "more pressing" things to do. Hopefully I should be able to get back to it this week-end or next week.

**Comment by fasterthanlime** [comment]
Caught this in real-world code yesterday: 

https://github.com/hapsoc/hring/blob/a2ffb643447b1e4bcdb673f3837b67cf3e02a19c/crates/hring/src/h1/client.rs#L34

(Not a showstopper, just thought I'd add a data point!)

**Comment by GuillaumeGomez** [comment]
I really need to find time to fix this...

**Comment by fasterthanlime** [comment]
> I really need to find time to fix this...

@GuillaumeGomez is this a good first issue? Is there a place one could get guidance to fix this?

**Comment by GuillaumeGomez** [comment]
Honestly I don't think it's a good first issue at all. You can see the code [here](https://github.com/rust-lang/rust-clippy/blob/master/clippy_lints/src/needless_pass_by_ref_mut.rs). If you want to start contributing on clippy, I'll be more than happy to help getting started but in this case, the problem is that the code generated by rustc for async is quite tricky to handle, especially variables being passed through which change IDs etc. Lots of fun. :3

**Comment by GuillaumeGomez** [comment]
I opened #11492 which fixes this issue.
